### PR TITLE
added changes for read recipients

### PR DESCRIPTION
**Pull request by Prithvi-MSFT**

### CODE REVIEW

File: samples/bot-conversation/csharp/Bots/TeamsConversationBot.cs [hunk 1]
```diff
@@ -72,6 +80,44 @@ protected override async Task OnTeamsMembersAddedAsync(IList<TeamsChannelAccount
             }
         }
 
+        private async Task CheckReadUserCount(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
+        {
+            if (teamMemberDetails.Count != 0)
+            {
+                int count = 0;
+                List<string> users = new List<string>();
+                foreach (var messageDetail in teamMemberReadDetails)
```
Comment (ChetanSharma-msft) [inline samples/bot-conversation/csharp/Bots/TeamsConversationBot.cs:89]: messageDetail  => memberDetails

File: samples/bot-conversation/csharp/README.md [hunk 2]
```diff
@@ -167,14 +167,43 @@ You can interact with this bot in Teams by sending it a message, or selecting a
...
+4. **Check Read count**
+  - **Result:** The bot will check the count of members who have read your message which sent through `MessageAllMembers`. User can also reset the count using `Reset read count` command
+  - **Valid Scopes:** personal, group chat, team chat
+
+  - **Personal Scope Interactions:**
+
+   **Check Read count command interaction:**
+  ![personal-CheckReadCount-interaction2 ](Images/personal-CheckReadCount-interaction2.png)
+
+   **Reset Read count command interaction:**
```
Comment (ChetanSharma-msft) [inline samples/bot-conversation/csharp/README.md:186]: Also, please update root readme contents that it support read receipt as well.

File: samples/bot-conversation/csharp/README.md [hunk 3]
```diff
@@ -210,5 +239,6 @@ To learn more about deploying a bot to Azure, see [Deploy your bot to Azure](htt
 - [Azure Bot Service Introduction](https://docs.microsoft.com/azure/bot-service/bot-service-overview-introduction?view=azure-bot-service-4.0)
 - [Azure Bot Service Documentation](https://docs.microsoft.com/azure/bot-service/?view=azure-bot-service-4.0)
 - [Messages in bot conversations](https://learn.microsoft.com/microsoftteams/platform/bots/how-to/conversations/conversation-messages?tabs=dotnet)
+- [Receive a read receipt](https://review.learn.microsoft.com/en-us/microsoftteams/platform/bots/how-to/conversations/conversation-messages?branch=pr-en-us-9184&tabs=dotnet1%2Capp-manifest-v112-or-later%2Cdotnet2%2Cdotnet3%2Cdotnet4%2Cdotnet5%2Cdotnet#receive-a-read-receipt)
```
Comment (ChetanSharma-msft) [inline samples/bot-conversation/csharp/README.md:242]: remove review.

File: samples/bot-conversation/csharp/README.md [hunk 4]
```diff
@@ -167,14 +167,43 @@ You can interact with this bot in Teams by sending it a message, or selecting a
...
+
+   **Check Read count command interaction:**
+  ![personal-CheckReadCount-interaction2 ](Images/personal-CheckReadCount-interaction2.png)
+
+   **Reset Read count command interaction:**
+  ![personal-ResetReadCount-interaction1 ](Images/personal-ResetReadCount-interaction1.png)
+
+   - **Group Chat Scope Interactions:**
+
+   **Check Read count command interaction:**
```
Comment (ChetanSharma-msft) [inline samples/bot-conversation/csharp/README.md:191]: Run build pipeline as well.

**Review by ChetanSharma-msft — CHANGES_REQUESTED** [review]
Added my review comments

File: samples/bot-conversation/csharp/README.md [hunk 5]
```diff
@@ -112,7 +112,7 @@ You can interact with this bot in Teams by sending it a message, or selecting a
   ![personal-AddedBot ](Images/personal-AddedBot.png)
```
[the remark this answers is not in the document]
Comment (Jegadeesh-MSFT) [reply samples/bot-conversation/csharp/README.md:112]: Existing Gif explained installation of app in all 3 scopes and shown **Mention me** functionality only in all scopes.  Whereas the existing code itself has nearly 6 functionalities and none was added there.

And so, I followed the same behavior.

File: samples/bot-conversation/csharp/Bots/TeamsConversationBot.cs [hunk 6]
```diff
@@ -72,6 +80,63 @@ protected override async Task OnTeamsMembersAddedAsync(IList<TeamsChannelAccount
...
+            teamMemberReadDetails = new ConcurrentDictionary<string, ChannelAccount>();
+        }
+
+        /// <summary>
+        /// Invoked when user read the message sent by bot in personal scope
+        /// </summary>
+        /// <param name="turnContext"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns>A task that represents the work queued to execute.</returns>
+        protected override async Task OnTeamsReadReceiptAsync(ReadReceiptInfo readReceiptInfo, ITurnContext<IEventActivity> turnContext, CancellationToken cancellationToken)
```
Comment (YunnyChung) [inline samples/bot-conversation/csharp/Bots/TeamsConversationBot.cs:133]: Can we show the example of using IsMessageRead? 

On the bot side, the bot developer will need to store one id of a bot's message (the message that a developer cares). ex: A company communicator bot sends an announcement to 5 people in the tenant. It wants to see whether everyone read that message. After the bot sent the announcmenet message, save that message id .
And whenver it receives read receipt event, compare the saved message id and ReadReceiptInfo.lastReadMessageId

File: samples/bot-conversation/csharp/Bots/TeamsConversationBot.cs [hunk 7]
```diff
@@ -72,6 +80,44 @@ protected override async Task OnTeamsMembersAddedAsync(IList<TeamsChannelAccount
             }
         }
 
+        private async Task CheckReadUserCount(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
```
[the remark this answers is not in the document]
Comment (Jegadeesh-MSFT) [reply samples/bot-conversation/csharp/Bots/TeamsConversationBot.cs:89]: Fixed

File: samples/bot-conversation/csharp/Bots/TeamsConversationBot.cs (hunk 1 above)
Comment (Jegadeesh-MSFT) [reply samples/bot-conversation/csharp/Bots/TeamsConversationBot.cs]: Fixed

File: samples/bot-conversation/csharp/Bots/TeamsConversationBot.cs [hunk 8]
```diff
@@ -72,6 +80,44 @@ protected override async Task OnTeamsMembersAddedAsync(IList<TeamsChannelAccount
...
+                foreach (var messageDetail in teamMemberReadDetails)
+                {
+                    if (teamMemberDetails.ContainsKey(messageDetail.Key))
+                    {
+                        count++;
+                        teamMemberDetails.TryGetValue(messageDetail.Key, out TeamsChannelAccount value);
+                        users.Add(value.Name);
+                    }
+                }
+                var userList = string.Join(", ", users);
```
[the remark this answers is not in the document]
Comment (Jegadeesh-MSFT) [reply samples/bot-conversation/csharp/Bots/TeamsConversationBot.cs:105]: Fixed

File: samples/bot-conversation/csharp/Bots/TeamsConversationBot.cs [hunk 9]
```diff
@@ -72,6 +80,44 @@ protected override async Task OnTeamsMembersAddedAsync(IList<TeamsChannelAccount
...
+                await turnContext.SendActivityAsync(MessageFactory.Text($"Number of members read the message : {count} \n\n Members : {userList}"), cancellationToken);
+            }
+            else
+            {
+                await turnContext.SendActivityAsync(MessageFactory.Text("Read count is zero. Please make sure to send message to all members firstly to check the count of members who have read your message."), cancellationToken);
+            }
+
+        }
+
+        private async Task ResetReadUserCount(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
```
[the remark this answers is not in the document]
Comment (Jegadeesh-MSFT) [reply samples/bot-conversation/csharp/Bots/TeamsConversationBot.cs:121]: Fixed

File: samples/bot-conversation/csharp/Bots/TeamsConversationBot.cs (hunk 6 above)
Comment (ChetanSharma-msft) [reply samples/bot-conversation/csharp/Bots/TeamsConversationBot.cs:123]: Implemented this
After the bot sent the announcmenet message, save that message id .
And whenver it receives read receipt event, compare the saved message id and ReadReceiptInfo.lastReadMessageId

File: samples/bot-conversation/csharp/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (ChetanSharma-msft) [reply samples/bot-conversation/csharp/README.md:112]: Please add .gif with all the functionality.

**Review by ChetanSharma-msft — CHANGES_REQUESTED** [review]
Added review comments

File: samples/bot-conversation/csharp/README.md [hunk 10]
```diff
@@ -167,14 +167,43 @@ You can interact with this bot in Teams by sending it a message, or selecting a
...
-  ![groupChat-BotCommands-interactions ](Images/groupChat-BotCommands-interactions.png)
+  ![groupChat-MessageAllMembers-interaction ](Images/groupChat-MessageAllMembers-interaction.png)
 
   - **Team Scope Interactions:**
 
    **MessageAllMembers command interaction:**
-  ![team-MessageAllMembers-interaction ](Images/team-MessageAllMembers-interaction.png)
+  ![team-MessageAllMembers-interactions ](Images/team-MessageAllMembers-interactions.png)
+
+4. **Check Read count**
```
[the remark this answers is not in the document]
Comment (Jegadeesh-MSFT) [reply samples/bot-conversation/csharp/README.md]: Fixed

File: samples/bot-conversation/csharp/README.md (hunk 3 above)
Comment (Jegadeesh-MSFT) [reply samples/bot-conversation/csharp/README.md]: Fixed

File: samples/bot-conversation/csharp/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (Jegadeesh-MSFT) [reply samples/bot-conversation/csharp/README.md:112]: Fixed

**Comment by Jegadeesh-MSFT** [comment]
**Executed Build pipeline**
https://github.com/OfficeDev/Microsoft-Teams-Samples/actions/runs/6182311703
image

File: samples/bot-conversation/csharp/README.md (hunk 4 above)
Comment (Jegadeesh-MSFT) [reply samples/bot-conversation/csharp/README.md:193]: Fixed

File: samples/bot-conversation/csharp/Bots/TeamsConversationBot.cs (hunk 6 above)
Comment (YunnyChung) [reply samples/bot-conversation/csharp/Bots/TeamsConversationBot.cs:123]: THank you!

File: samples/bot-conversation/csharp/Images/groupChat-CheckReadCount-interaction.png
Comment (YunnyChung) [inline samples/bot-conversation/csharp/Images/groupChat-CheckReadCount-interaction.png:1]: Hmm I am confused. The read receipt event for the bot messages is only supported in personal scope. Does it work in group chat?

File: samples/bot-conversation/csharp/Images/team-CheckReadCount-interaction.png
Comment (YunnyChung) [inline samples/bot-conversation/csharp/Images/team-CheckReadCount-interaction.png:1]: Same here. Read Receipt event feature for bot messages is only supported in personal scope. Does the bot receive read receipt event when a user reads the bot's message sent in group chat or team/channel?

File: README.md [hunk 11]
```diff
@@ -63,7 +63,7 @@ The [Teams Toolkit](https://marketplace.visualstudio.com/items?itemName=TeamsDev
...
 |1| Teams Conversation Bot quick-start | Messaging and conversation event handling hello world. | | [View][bot-conversation-quickstart#js] ![toolkit-icon](assets/toolkit-icon.png) | | |
 |2| Teams Conversation Bot SSO quick-start | Messaging and conversation event handling hello world with SSO. | [View][bot-conversation-sso-quickstart#csharp_dotnetcore] | [View][bot-conversation-sso-quickstart#js] ![toolkit-icon](assets/toolkit-icon.png) | | |
-|3| Teams Conversation Bot | Messaging and conversation event handling. | [View][bot-conversation#cs]| [View][bot-conversation#js] ![toolkit-icon](assets/toolkit-icon.png) | [View][bot-conversation#python] | [View][bot-conversation#java] |[View](/samples/bot-conversation/csharp/demo-manifest/bot-conversation.zip)
+|3| Teams Conversation Bot | Messaging and conversation event handling also includes read receipt functionality. | [View][bot-conversation#cs]| [View][bot-conversation#js] ![toolkit-icon](assets/toolkit-icon.png) | [View][bot-conversation#python] | [View][bot-conversation#java] |[View](/samples/bot-conversation/csharp/demo-manifest/bot-conversation.zip)
```
Comment (YunnyChung) [inline README.md:66]: If this Teams Conversation Bot is supported in all scopes, let's explicitly mention that this bot handles read receipt event in personal scope. ex :"includes read receipt event functionality in personal scope"

File: samples/bot-conversation/csharp/Images/groupChat-CheckReadCount-interaction.png
Comment (Jegadeesh-MSFT) [reply samples/bot-conversation/csharp/Images/groupChat-CheckReadCount-interaction.png:1]: Hi @YunnyChung ,

Read receipt functionality only supports personal scope. In group chats, messages are sent to each individual's personal scope, and we can count them when they view those messages in their personal scope.

File: samples/bot-conversation/csharp/Images/team-CheckReadCount-interaction.png
Comment (Jegadeesh-MSFT) [reply samples/bot-conversation/csharp/Images/team-CheckReadCount-interaction.png:1]: no, we are sending messages to all members of the group chat to their personal scope. So, when a person views that message in their personal scope we are able to get the name and count through read receipt functionality.

File: README.md (hunk 11 above)
Comment (Jegadeesh-MSFT) [reply README.md]: Fixed

**Review by ChetanSharma-msft — APPROVED** [review]
Changes looks fine, approving it.

File: samples/bot-conversation/csharp/Bots/TeamsConversationBot.cs (hunk 6 above)
Comment (ChetanSharma-msft) [reply samples/bot-conversation/csharp/Bots/TeamsConversationBot.cs:123]: @Jegadeesh-MSFT - Please confirm about this functionality: **IsMessageRead** and implement it accordingly.

**Comment by ChetanSharma-msft** [comment]
Executed build pipeline: https://github.com/OfficeDev/Microsoft-Teams-Samples/actions/runs/6217900710

**Comment by Jegadeesh-MSFT** [comment]
**Executed Build pipeline**

[Build - All Samples · OfficeDev/Microsoft-Teams-Samples@d60a635 (github.com)](https://github.com/OfficeDev/Microsoft-Teams-Samples/actions/runs/6218499429)

image

**Review by YunnyChung — APPROVED** [review]

**Comment by Jegadeesh-MSFT** [comment]
Created new PR due to merge conflicts.

New PR : https://github.com/OfficeDev/Microsoft-Teams-Samples/pull/973

**Comment by ChetanSharma-msft** [comment]
Duplicate of: https://github.com/OfficeDev/Microsoft-Teams-Samples/pull/973

**Pull request closed without merging**